udtube/callbacks.py

```diff
@@ -49,38 +49,47 @@ def write_on_batch_end(
+        upos_hat = (
             torch.argmax(logits.upos, dim=1) if logits.use_upos else None
         )
-        y_xpos_hat = (
+        xpos_hat = (
             torch.argmax(logits.xpos, dim=1) if logits.use_xpos else None
         )
-        y_lemma_hat = (
+        lemma_hat = (
             torch.argmax(logits.lemma, dim=1) if logits.use_lemma else None
         )
-        y_feats_hat = (
+        feats_hat = (
             torch.argmax(logits.feats, dim=1) if logits.use_feats else None
         )
         for i, tokenlist in enumerate(batch.tokenlists):
-            # Edits the relevant fields in each tokenlist.
-            if y_upos_hat is not None:
-                upos_hat = self.mapper.decode_upos(y_upos_hat[i, :])
-                for j, upos in enumerate(upos_hat[: len(tokenlist)]):
-                    tokenlist[j]["upos"] = upos
-            if y_xpos_hat is not None:
-                xpos_hat = self.mapper.decode_xpos(y_xpos_hat[i, :])
-                for j, xpos in enumerate(xpos_hat[: len(tokenlist)]):
-                    tokenlist[j]["xpos"] = xpos
-            if y_lemma_hat is not None:
-                lemma_hat = self.mapper.decode_lemma(
-                    [token["form"] for token in tokenlist],
-                    y_lemma_hat[i, :],
+            # Sentence-level decoding of the classification indices, followed
+            # by rewriting the fields in the tokenlist.
+            if upos_hat is not None:
+                upos_it = self.mapper.decode_upos(upos_hat[i, :])
+                for token in tokenlist:
+                    if token.is_mwe:
+                        continue
+                    token.upos = next(upos_it)
+            if xpos_hat is not None:
+                xpos_it = self.mapper.decode_xpos(xpos_hat[i, :])
+                for token in tokenlist:
+                    if token.is_mwe:
+                        continue
+                    token.xpos = next(xpos_it)
+            if lemma_hat is not None:
+                lemma_it = self.mapper.decode_lemma(
+                    tokenlist.get_tokens(), lemma_hat[i, :]
                 )
-                for j, lemma in enumerate(lemma_hat[: len(tokenlist)]):
-                    tokenlist[j]["lemma"] = lemma
-            if y_feats_hat is not None:
-                feats_hat = self.mapper.decode_feats(y_feats_hat[i, :])
-                for j, feats in enumerate(feats_hat[: len(tokenlist)]):
-                    tokenlist[j]["feats"] = feats
-            print(tokenlist.serialize(), file=self.sink)
+                for token in tokenlist:
+                    if token.is_mwe:
+                        continue
+                    token.lemma = next(lemma_it)
+            if feats_hat is not None:
+                feats_it = self.mapper.decode_feats(feats_hat[i, :])
+                for token in tokenlist:
+                    if token.is_mwe:
+                        continue
+                    token.feats = next(feats_it)
```

Comment:
This has to be possible to put into a helper function in some way. Something like

```
def _good_name(it_tracker, class_attr):
    for token in tokenlist:
                    if token.is_mwe:
                        continue
                    class_attr = next(it_tracker)
```

Reply:
Obviously variable naming and all needs work.

Reply:
I was skeptical but I did it and it's pretty good, take a look.